src/drivers/nvme.c

```diff
@@ -0,0 +1,865 @@
+    cmd->dptr.prp1 = physical_from_virtual(n->iosq.ring);
+    cmd->cdw10 = (MASK(n->ioq_order) << 16) | NVME_IOQ_IDX; /* queue size and queue ID */
+    cmd->cdw11 = (NVME_IOQ_IDX << 16) | 0x01;  /* completion queue ID, physically contiguous */
+    nvme_sq_doorbell(n, NVME_AQ_IDX, &n->asq);
+    return true;
+}
+
+closure_function(2, 0, void, nvme_create_iocq_resp,
+                 nvme, n, storage_attach, a)
+{
+    nvme n = bound(n);
+    struct nvme_cqe *cqe = nvme_get_cqe(&n->acq);
+    if (cqe) {
+        n->asq.head = NVME_SQ_HEAD(cqe->dw2);
+        int sc = NVME_STATUS_CODE(cqe->dw3);
+        nvme_cq_doorbell(n, NVME_AQ_IDX, &n->acq);
+        if (sc == NVME_SC_OK) {
+            nvme_debug("I/O CQ created");
+            nvme_create_iosq(n, bound(a));
+        } else {
+            msg_err("failed to create I/O CQ: status code 0x%x\n", sc);
+        }
+    }
+    closure_finish();
+}
+
+static boolean nvme_create_iocq(nvme n, storage_attach a)
+{
+    if (!nvme_init_cq(n, &n->iocq, n->ioq_order)) {
+        msg_err("failed to initialize queue\n");
+        return false;
+    }
+    n->ac_handler = closure(n->general, nvme_create_iocq_resp, n, a);
+    if (n->ac_handler == INVALID_ADDRESS) {
+        msg_err("failed to allocate completion handler\n");
+        nvme_deinit_cq(n, &n->iocq);
+        return false;
+    }
+    pci_setup_msix(n->d, NVME_IOQ_MSIX, init_closure(&n->io_irq, nvme_io_irq, n), "nvme I/O");
+    struct nvme_sqe *cmd = nvme_get_sqe(&n->asq);
+    zero(cmd, sizeof(*cmd));
+    cmd->cdw0 = NVME_CID(n->asq.tail) | NVME_CMD_PRP | NVME_OPC_CRE_IOCQ;
+    cmd->dptr.prp1 = physical_from_virtual(n->iocq.ring);
+    cmd->cdw10 = (MASK(n->ioq_order) << 16) | NVME_IOQ_IDX; /* queue size and queue ID */
+    cmd->cdw11 = (NVME_IOQ_MSIX << 16) | 0x03;  /* interrupts enabled, physically contiguous */
+    nvme_sq_doorbell(n, NVME_AQ_IDX, &n->asq);
+    return true;
+}
+
+define_closure_function(1, 0, void, nvme_admin_irq,
+                        nvme, n)
+{
+    nvme n = bound(n);
+    nvme_debug("%s (%F)", __func__, n->ac_handler);
+    apply(n->ac_handler);
+}
+
+closure_function(3, 1, boolean, nvme_probe,
+                 heap, general, storage_attach, a, heap, contiguous,
+                 pci_dev, d)
+{
+    if ((pci_get_class(d) != PCIC_STORAGE) || (pci_get_subclass(d) != PCIS_STORAGE_NVM) ||
+            (pci_get_prog_if(d) != PCIPI_STORAGE_NVME))
+        return false;
+    heap general = bound(general);
+    nvme n = allocate(general, sizeof(*n));
+    if (n == INVALID_ADDRESS)
+        return false;
+    n->general = general;
+    n->contiguous = bound(contiguous);
+    if (!nvme_init_sq(n, &n->asq, NVME_ASQ_ORDER))
+        goto free_nvme;
+    if (!nvme_init_cq(n, &n->acq, NVME_ACQ_ORDER))
+        goto deinit_asq;
+    pci_bar_init(d, &n->bar, NVME_REG_BAR, 0, -1);
+
+    /* reset controller */
+    pci_bar_write_4(&n->bar, NVME_CC, 0);
+    while (pci_bar_read_4(&n->bar, NVME_CSTS) & NVME_CSTS_RDY)
```

Comment:
Not sure if this is really necessary for a virtualized device, but since this is a probe function - maybe these ready waits could error on a timeout?

Reply:
I added timeout handling here and also a few lines below (where the controller is enabled).